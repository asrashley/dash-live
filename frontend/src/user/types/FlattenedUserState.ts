import type { InitialUserState } from "../types/InitialUserState";


export type FlattenedUserState = Omit<InitialUserState, 'groups'> & {
  adminGroup: boolean;
  mediaGroup: boolean;
  userGroup: boolean;
};
