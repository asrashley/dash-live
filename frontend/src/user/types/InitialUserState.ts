export type InitialUserState = {
    pk?: number;
    username?: string;
    email?: string;
    lastLogin: string | null;
    mustChange: boolean;
    groups: string[];
};
