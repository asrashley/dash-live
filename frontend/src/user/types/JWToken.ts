export type JWToken = {
  expires: string;
  jwt: string;
};
