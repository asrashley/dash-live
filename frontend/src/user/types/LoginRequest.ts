export interface LoginRequest {
    username: string;
    password: string;
    rememberme: boolean;
}