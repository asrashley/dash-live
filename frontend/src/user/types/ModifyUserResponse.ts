import type { InitialUserState } from "./InitialUserState";

export interface ModifyUserResponse {
    success: boolean;
    errors: string[];
    user?: InitialUserState;
}
