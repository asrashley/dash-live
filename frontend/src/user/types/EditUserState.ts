import { FlattenedUserState } from "./FlattenedUserState";


export type EditUserState = FlattenedUserState & {
  password?: string;
  confirmPassword?: string;
};
