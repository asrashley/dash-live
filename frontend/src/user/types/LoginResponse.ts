import { JWToken } from "./JWToken";
import { InitialUserState } from "./InitialUserState";

export interface LoginResponse {
    success: boolean;
    error?: string;
    csrf_token: string;
    accessToken?: JWToken;
    refreshToken?: JWToken;
    user?: InitialUserState
}
