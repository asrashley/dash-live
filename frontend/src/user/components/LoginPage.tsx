import { useCallback, useContext } from "preact/hooks";
import { useSignal } from "@preact/signals";
import { useLocation } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";

import { EndpointContext } from "../../endpoints";
import { LoginRequest } from "../types/LoginRequest";
import { LoginResponse } from "../types/LoginResponse";
import { WhoAmIContext } from "../hooks/useWhoAmI";
import { useLocalStorage } from "../../hooks/useLocalStorage";

import { LoginCard } from "./LoginCard";

export default function LoginPage() {
  const apiRequests = useContext(EndpointContext);
  const { setUser } = useContext(WhoAmIContext);
  const { setRefreshToken } = useLocalStorage();
  const setLocation = useLocation()[1];
  const error = useSignal<string | undefined>();
  const submitting = useSignal<boolean>(false);
  const onLogin = useCallback((request: LoginRequest) => {
    submitting.value = true;
    apiRequests.loginUser(request).then((resp: LoginResponse) => {
        if (resp.success) {
            error.value = undefined;
            setUser(resp.user);
            setRefreshToken(resp.refreshToken ?? null);
            setLocation(uiRouteMap.home.url());
        } else {
            error.value = resp.error ?? 'Unknown error';
        }
    }).catch(err => {
        error.value = `Login failed - ${err}`;
    }).finally(() =>{
        submitting.value = false;
    });
  }, [apiRequests, error, setLocation, setRefreshToken, setUser, submitting]);

  return <LoginCard submitting={submitting} error={error} onLogin={onLogin} />;
}
