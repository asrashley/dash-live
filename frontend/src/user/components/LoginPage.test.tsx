import { describe, expect, test } from "vitest";
import userEvent from "@testing-library/user-event";
import { mock } from "vitest-mock-extended";

import { renderWithProviders } from "../../test/renderWithProviders";
import { ApiRequests, EndpointContext } from "../../endpoints";
import LoginPage from "./LoginPage";
import { LoginResponse } from "../../types/LoginResponse";
import { LoginRequest } from "../../types/LoginRequest";

describe("LoginPage component", () => {
  const apiRequests = mock<ApiRequests>();

  test("matches snapshot", () => {
    const { asFragment } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <LoginPage />
      </EndpointContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('successful login', async () => {
    const loginResponse: LoginResponse = {
        success: true,
        accessToken: {
            jwt: 'access.abc123',
            expires: '2025-01-09T22:33:44Z',
        },
        refreshToken: {
            jwt: 'refresh.abc123',
            expires: '2025-01-09T22:33:44Z',
        },
        user: {
            isAuthenticated: true,
            pk: 42,
            username: 'my.username',
            email: 'unit.test@local',
            groups: ['USERS'],
        },
        csrf_token: "csrf.7654"
    };
    const loginPromise = new Promise<LoginRequest>((resolve => {
        apiRequests.loginUser.mockImplementation((req: LoginRequest) => {
            resolve(req);
            return Promise.resolve(loginResponse);
        });
    }))
    const user = userEvent.setup();
    const { getByLabelText, getByText } = renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <LoginPage />
        </EndpointContext.Provider>
      );
      const userInp = getByLabelText("Username", {
        exact: false,
      }) as HTMLInputElement;
      const pwdInp = getByLabelText("Password", {
        exact: false,
      }) as HTMLInputElement;
      await user.type(userInp, "test");
      await user.type(pwdInp, "secret");
      const submit = getByText('Login') as HTMLButtonElement;
      expect(submit.disabled).toEqual(false);
      await user.click(submit);
      await expect(loginPromise).resolves.toEqual({
        username: 'test',
        password: 'secret',
        rememberme: false,
      });
  });

  test('failed login', async () => {
    const loginResponse: LoginResponse = {
        success: false,
        error: 'Incorrect username or password',
        csrf_token: "csrf.7654"
    };
    const loginPromise = new Promise<void>((resolve => {
        apiRequests.loginUser.mockImplementation(() => {
            resolve();
            return Promise.resolve(loginResponse);
        });
    }))
    const user = userEvent.setup();
    const { getByLabelText, getBySelector, getByText, findByText } = renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <LoginPage />
        </EndpointContext.Provider>
      );
      const userInp = getByLabelText("Username", {
        exact: false,
      }) as HTMLInputElement;
      const pwdInp = getByLabelText("Password", {
        exact: false,
      }) as HTMLInputElement;
      await user.type(userInp, "test");
      await user.type(pwdInp, "secret");
      const submit = getByText('Login') as HTMLButtonElement;
      expect(submit.disabled).toEqual(false);
      await user.click(submit);
      await loginPromise;
      await findByText('Incorrect username or password');
      expect((getBySelector("#login-form") as HTMLElement).className).toEqual("");
  });

  test('fails to make login API call', async () => {
    const loginPromise = new Promise<void>((resolve => {
        apiRequests.loginUser.mockImplementation(() => {
            resolve();
            return Promise.reject(new Error('NetworkError when attempting to fetch resource'));
        });
    }));
    const user = userEvent.setup();
    const { getByLabelText, getBySelector, getByText, findByText } = renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <LoginPage />
        </EndpointContext.Provider>
      );
      const userInp = getByLabelText("Username", {
        exact: false,
      }) as HTMLInputElement;
      const pwdInp = getByLabelText("Password", {
        exact: false,
      }) as HTMLInputElement;
      await user.type(userInp, "test");
      await user.type(pwdInp, "secret");
      const submit = getByText('Login') as HTMLButtonElement;
      expect(submit.disabled).toEqual(false);
      await user.click(submit);
      await loginPromise;
      await findByText('NetworkError when attempting to fetch resource', { exact: false });
      expect((getBySelector("#login-form") as HTMLElement).className).toEqual("");
  });
});
