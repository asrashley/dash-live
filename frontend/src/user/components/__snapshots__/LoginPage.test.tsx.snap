// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card "
    id="login"
  >
    <div
      class="card-header"
    >
      Log into DASH server
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
        id="login-form"
        name="login"
      >
        <div
          class="d-none"
          role="alert"
        />
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-username"
            id="label-username"
          >
            Username:
          </label>
          <div
            class="col-10"
          >
            <input
              aria-describedby="label-username"
              class="form-control "
              id="model-username"
              name="username"
              title="Username"
              type="text"
            />
          </div>
          <div
            class="invalid-feedback d-none"
          />
        </div>
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-password"
            id="label-password"
          >
            Password:
          </label>
          <div
            class="col-10"
          >
            <div
              class="input-group"
            >
              <input
                aria-describedby="label-password"
                class="form-control "
                id="model-password"
                name="password"
                title="Password"
                type="password"
              />
              <button
                class="d-none"
                type="button"
              >
                Reveal
              </button>
            </div>
          </div>
          <div
            class="invalid-feedback d-none"
          />
        </div>
        <div
          class="row mb-2 form-group "
        >
          <label
            class="col-2 col-form-label"
            for="model-rememberme"
            id="label-rememberme"
          >
            Remember Me:
          </label>
          <div
            class="col-10"
          >
            <input
              aria-describedby="label-rememberme"
              class="form-check-input "
              id="model-rememberme"
              name="rememberme"
              title="Remember Me"
              type="checkbox"
            />
          </div>
          <div
            class="invalid-feedback d-none"
          />
        </div>
        <div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
